Bottom: 973826849f1c5d5dde2b046b551bd9a02f337a65
Top:    1bc1dc7fe92a356b0e4fe0e027f5e8b69e2c6672
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2015-01-01 17:10:53 +0000

Add hysteresis to grid-snapping


---

diff --git a/src/crosshair.c b/src/crosshair.c
index 73ec90a..ee57d6b 100644
--- a/src/crosshair.c
+++ b/src/crosshair.c
@@ -929,6 +929,11 @@ FitCrosshairIntoGrid (Coord X, Coord Y)
   struct snap_data snap_data;
   int ans;
 
+  Coord old_x, old_y;
+
+  old_x = Crosshair.X;
+  old_y = Crosshair.Y;
+
   Crosshair.X = CLAMP (X, Crosshair.MinX, Crosshair.MaxX);
   Crosshair.Y = CLAMP (Y, Crosshair.MinY, Crosshair.MaxY);
 
@@ -955,11 +960,26 @@ FitCrosshairIntoGrid (Coord X, Coord Y)
     }
 
   snap_data.crosshair = &Crosshair;
-  snap_data.nearest_sq_dist =
-    crosshair_sq_dist (&Crosshair, nearest_grid_x, nearest_grid_y);
   snap_data.nearest_is_grid = true;
   snap_data.x = nearest_grid_x;
   snap_data.y = nearest_grid_y;
+#if 0
+  snap_data.nearest_sq_dist = crosshair_sq_dist (&Crosshair, snap_data.x, snap_data.y);
+
+  if (snap_data.nearest_sq_dist > PCB->Grid / 3 * PCB->Grid / 3)
+    {
+      snap_data.x = old_x;
+      snap_data.y = old_y;
+    }
+#endif
+
+  if (labs (nearest_grid_x - Crosshair.X) > PCB->Grid / 3)
+    snap_data.x = old_x;
+
+  if (labs (nearest_grid_y - Crosshair.Y) > PCB->Grid / 3)
+    snap_data.y = old_y;
+
+  snap_data.nearest_sq_dist = crosshair_sq_dist (&Crosshair, snap_data.x, snap_data.y);
 
   ans = NO_TYPE;
   if (!PCB->RatDraw)
